src/v/cluster/rm_stm.cc

```diff
@@ -790,19 +777,6 @@ ss::future<tx_errc> rm_stm::do_abort_tx(
       pid,
       tx_seq.value_or(model::tx_seq(-1)),
       synced_term);
-    // catching up with all previous end_tx operations (commit | abort)
-    // to avoid writing the same commit | abort marker twice
-    if (_mem_state.last_end_tx >= model::offset{0}) {
-        if (!co_await wait_no_throw(
-              _mem_state.last_end_tx, model::timeout_clock::now() + timeout)) {
-            vlog(
-              _ctx_log.trace,
-              "Can't catch up to abort pid:{} tx_seq:{}",
-              pid,
-              tx_seq.value_or(model::tx_seq(-1)));
-            co_return tx_errc::stale;
-        }
-    }
```

Comment:
I am wondering if it is possible that the abort or commit control batch was replicated but not yet applied we update the `_mem_state.last_end_tx` after call to `raft::replicate` is finished but before we wait for the offset to be applied. I see that there is locking in place i.e. we grab `tx_lock` whenever something is being done with producer transactions. In this case the wait is global  i.e. it provides ordering between different producers. I do not know if that is desired and required, this is just an observation. If this ordering between producers is not relevant this is LGTM

Reply:
AFAICT the ordering between producers doesn't matter because the state is segmented by producer, a producer and it's transaction lifecycle only cares about its state. The shared state is mostly tracking LSO, aborted transactions etc but those are not affected by this situation IMO.